Fix resource.MustParse fails to parse quantities near math.MaxInt64

dwbcz opened the pull request:
#### What type of PR is this?

/kind bug

#### What this PR does / why we need it:

The current checks using `maxInt64Factors` are too coarse: they allow `999999999999999999 (has 18 nines)` to pass but block `1000000000000000000 (has 18 0)` 
```go
largestPass := "999999999999999999"
q := resource.MustParse(largestPass)
fmt.Println(q.AsInt64())

smallestBlock := "1000000000000000000"
q2 := resource.MustParse(smallestBlock)
fmt.Println(q2.AsInt64())
fmt.Println(q2.AsDec())
```
#### output
```
999999999999999999 true
0 false
1000000000000000000.000000000
```
 This pr added a precise boundary check so that up to `9223372036854775807` (maxInt64) can be accepted, while values above` 9223372036854775807` up to `999999999999999999 (has 19 nines)` are still correctly rejected. 
previous unit tests didn’t cover large int case before, this pr added some test cases for it as well.

#### Which issue(s) this PR is related to:
Fixes #135487
<https://github.com/kubernetes/kubernetes/issues/135487>

#### Special notes for your reviewer:

Compared to https://github.com/kubernetes/kubernetes/pull/135595, this PR explores a different approach to fixing the parsing behavior.

#### Does this PR introduce a user-facing change?
```release-note
Fix parsing of very large quantity values near `math.MaxInt64` in `resource.MustParse`, preventing incorrect parsing results and ensuring consistency with `resource.NewQuantity`.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
```docs
NONE
```

dwbcz committed:
Fix resource.MustParse fails to parse quantities near math.MaxInt64

Previously, the parser used a simple heuristic: if the concatenated digits (num+denom) exceed 18, it falls back to the Dec (arbitrary precision) path. While safe, this incorrectly treats some 19-digit values that still fit in int64 (≤ MaxInt64) as non-representable by the fast path.

This change refines the heuristic:

Keep the 18-digit fast-path rule.
When the total digits are exactly 19, perform an explicit boundary check against MaxInt64 using the concatenated numeric string.
If the value fits (≤ MaxInt64), allow the fast path by treating precision as non-negative.
Result:

Correctly handles all representable values, including 19-digit numbers like 10^18, without unnecessarily falling back to Dec.
Retains safety for values that exceed MaxInt64.

carlory commented:
@nathannewyen has submitted a [PR](https://github.com/kubernetes/kubernetes/pull/135595) to fix the issue https://github.com/kubernetes/kubernetes/issues/135487
Could you help review the PR?

dwbcz committed:
This is a regression test for issue #135487 where resource.MustParse
fails to parse quantities near math.MaxInt64 correctly, value larger
than 1000000000000000000(1e18) previously failed because maxInt64Factors = 18
after fix it can pass this test

nathannewyen commented:
> > @nathannewyen has submitted a [PR](https://github.com/kubernetes/kubernetes/pull/135595) to fix the issue #135487 Could you help review the PR?
> 
> Hi, Thanks for review!! I’ve reviewed that PR. I’m taking a different approach: that PR applies a patch around another function, while mine addresses the root cause at its source. I’ll add more details on the issue page. In fact, the two PRs don’t conflict and can both be merged.

Thanks for the clarification @dwbcz!

siyuanfoundation reviewed (commented):
this solution looks reasonable to me. 
One minor comment.

dwbcz committed:
add more testcase for fix of #135487

dwbcz commented:
/retest

dwbcz committed:
refine unit test: set expectError field

dwbcz commented:
/retest

dwbcz commented on staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go:
```diff
@@ -553,6 +553,73 @@ func TestQuantityParse(t *testing.T) {
+		{"-922337203685477580", intQuantity(-922337203685477580, 0, DecimalSI), false},
+		// value larger than 1000000000000000000(1e18) would fail if skipping attempting
+		// to convert inf.Dec because maxInt64Factors = 18, see issue#135487
+		{"1000000000000000000", intQuantity(1000000000000000000, 0, DecimalSI), false},
```
Yes, this is expected. `TestQuantityParse` is initialized with `decQuantity`. So all the test case in `TestQuantityParse` is using dec path

dwbcz replied:
I added this log line in function `TestQuantityParse`, as a double-check. From the output, I can confirm that all test cases `continue` at this point, since they all follow the decimal (dec) path
<img width="1303" height="145" alt="image" src="https://github.com/user-attachments/assets/a9fc21bf-bc2c-4589-856f-a41a5977b885" />
➜  resource go test -v -run TestQuantityParse
go: downloading go1.25.0 (linux/amd64)
=== RUN   TestQuantityParseZero
--- PASS: TestQuantityParseZero (0.00s)
=== RUN   TestQuantityParseNonNumericPanic
--- PASS: TestQuantityParseNonNumericPanic (0.00s)
=== RUN   TestQuantityParse
    quantity_test.go:421: 0: expected inf.Dec to return true for AsInt64: {{0 0} {0xc000126d20}  DecimalSI}
    quantity_test.go:421: 0n: expected inf.Dec to return true for AsInt64: {{0 0} {0xc000126d50}  DecimalSI}
    quantity_test.go:421: 0u: expected inf.Dec to return true for AsInt64: {{0 0} {0xc000126d80}  DecimalSI}
    quantity_test.go:421: 0m: expected inf.Dec to return true for AsInt64: {{0 0} {0xc000126db0}  DecimalSI}
    ........

ivange94 commented on staging/src/k8s.io/apimachinery/pkg/api/resource/quantity_test.go:
```diff
@@ -553,6 +553,73 @@ func TestQuantityParse(t *testing.T) {
+		if !ok {
+			continue
+		}
+		if !ok {
```
Looks like the code in this block will never be called due to the the `continue` on line 609? If `!ok` is true, then line 611 and below is ignored altogether. If it's false, then the block in this if is still ignored.

dwbcz committed:
refine unit test

thockin commented:
I count no less than 6 PRs for this.  Who is comparing impl options?

jpbetz commented:
> we have 8 PRs open trying to fix the same issue, @jpbetz will coordinate picking the one to proceed with in [#135487 (comment)](https://github.com/kubernetes/kubernetes/issues/135487#issuecomment-4215382624)

Thanks to everyone who has proposed contribute solutions to this issue. We apologize that non of the PRs opened to try to fix this issue did not get the attention they deserve.  I've had a look at all of them.

Since we can only merge one fix. We are working toward merging https://github.com/kubernetes/kubernetes/pull/138076.  I am closing all other PRs on this issue.

I encourage anyone interested in this to provide feedback on https://github.com/kubernetes/kubernetes/pull/138076 to ensure we get a high quality fix merged.

/close
